File: webapp/src/api/bundles.ts
import { fetchApi } from './client';

// Module reference in a bundle
export interface ModuleRef {
  module: string;
  source?: string | null;
  config?: Record<string, unknown> | null;
}

// Module reference with source tracking
export interface ResolvedModuleRef extends ModuleRef {
  definedIn: string;
  overridden: boolean;
  overrideIn?: string | null;
  originalConfig?: Record<string, unknown> | null;
}

// Session config reference
export interface SessionConfigRef {
  orchestrator?: ModuleRef | null;
  context?: ModuleRef | null;
}

// Resolved session config
export interface ResolvedSessionConfig {
  orchestrator?: ResolvedModuleRef | null;
  context?: ResolvedModuleRef | null;
}

// Bundle list item with summary
export interface BundleListItem {
  name: string;
  version: string;
  description?: string | null;
  source: 'user' | 'system';
  path: string;

  // Quick stats
  providerCount: number;
  toolCount: number;
  hookCount: number;
  agentCount: number;

  // Composition
  includes: string[];
}

// Full bundle details (raw structure)
export interface BundleDetails extends BundleListItem {
  session?: SessionConfigRef | null;
  providers: ModuleRef[];
  tools: ModuleRef[];
  hooks: ModuleRef[];
  agents: ModuleRef[];
  context: Record<string, string>;
  instruction?: string | null;
}

// Tree node for includes hierarchy
export interface IncludesTreeNode {
  name: string;
  includes: IncludesTreeNode[];
}

// Resolved bundle with source tracking
export interface ResolvedBundle {
  name: string;
  source: 'user' | 'system';
  gitUrl?: string | null;
  includesChain: string[];
  includesTree: IncludesTreeNode;

  session?: ResolvedSessionConfig | null;
  providers: ResolvedModuleRef[];
  tools: ResolvedModuleRef[];
  hooks: ResolvedModuleRef[];
  agents: ResolvedModuleRef[];

  instruction?: string | null;
}

// Raw bundle source
export interface BundleSource {
  name: string;
  content: string;
  path: string;
  format: 'md' | 'yaml' | 'directory';
}

// Request types
export interface CreateBundleRequest {
  name: string;
  baseBundle?: string;
  description?: string;
}

export interface CopyBundleRequest {
  newName: string;
}

export interface RenameBundleRequest {
  newName: string;
}

export interface AddRegistryBundleRequest {
  name: string;
  gitUrl: string;
}

export interface UpdateBundleRequest {
  content: string;
}

// API functions

// List all bundles with summary info
export const listBundles = () =>
  fetchApi<BundleListItem[]>('/api/v1/bundles/');

// Get full bundle details
export const getBundleDetails = (name: string) =>
  fetchApi<BundleDetails>(`/api/v1/bundles/${encodeURIComponent(name)}/`);

// Get resolved bundle with source tracking
export const getResolvedBundle = (name: string) =>
  fetchApi<ResolvedBundle>(`/api/v1/bundles/${encodeURIComponent(name)}/resolved`);

// Get raw bundle source
export const getBundleSource = (name: string) =>
  fetchApi<BundleSource>(`/api/v1/bundles/${encodeURIComponent(name)}/source`);

// Create a new bundle
export const createBundle = (data: CreateBundleRequest) =>
  fetchApi<BundleListItem>('/api/v1/bundles/', {
    method: 'POST',
    body: JSON.stringify(data),
  });

// Copy a bundle
export const copyBundle = (name: string, newName: string) =>
  fetchApi<BundleListItem>(`/api/v1/bundles/${encodeURIComponent(name)}/copy`, {
    method: 'POST',
    body: JSON.stringify({ newName } as CopyBundleRequest),
  });

// Rename a bundle
export const renameBundle = (name: string, newName: string) =>
  fetchApi<BundleListItem>(`/api/v1/bundles/${encodeURIComponent(name)}/rename`, {
    method: 'POST',
    body: JSON.stringify({ newName } as RenameBundleRequest),
  });

// Update a bundle
export const updateBundle = (name: string, content: string) =>
  fetchApi<BundleListItem>(`/api/v1/bundles/${encodeURIComponent(name)}/`, {
    method: 'PUT',
    body: JSON.stringify({ content } as UpdateBundleRequest),
  });

// Delete a bundle
export const deleteBundle = (name: string) =>
  fetchApi<{ message: string }>(`/api/v1/bundles/${encodeURIComponent(name)}/`, {
    method: 'DELETE',
  });

// Add a bundle from git URL to registry
export const addRegistryBundle = (name: string, gitUrl: string) =>
  fetchApi<{ message: string }>('/api/v1/bundles/registry', {
    method: 'POST',
    body: JSON.stringify({ name, gitUrl } as AddRegistryBundleRequest),
  });

// Remove a bundle from registry
export const removeRegistryBundle = (name: string) =>
  fetchApi<{ message: string }>(`/api/v1/bundles/registry/${encodeURIComponent(name)}`, {
    method: 'DELETE',
  });

File: lakehoused/lakehoused/models/bundles.py
"""API models for bundle operations."""

from pydantic import Field

from lakehoused.models.base import CamelCaseModel


class ModuleRef(CamelCaseModel):
    """Module reference in a bundle."""

    module: str = Field(description="Module identifier")
    source: str | None = Field(default=None, description="Module source URL or path")
    config: dict[str, object] | None = Field(default=None, description="Module configuration")


class ResolvedModuleRef(CamelCaseModel):
    """Module reference with source tracking for resolved bundles."""

    module: str = Field(description="Module identifier")
    source: str | None = Field(default=None, description="Module source URL or path")
    config: dict[str, object] | None = Field(default=None, description="Module configuration")

    # Source tracking
    defined_in: str = Field(description="Which bundle defined this module")
    overridden: bool = Field(default=False, description="Whether this is overridden in the current bundle")
    override_in: str | None = Field(default=None, description="Which bundle overrides this (if overridden)")
    original_config: dict[str, object] | None = Field(
        default=None, description="Original config before override (if overridden)"
    )


class SessionConfigRef(CamelCaseModel):
    """Session configuration reference."""

    orchestrator: ModuleRef | None = Field(default=None, description="Orchestrator module")
    context: ModuleRef | None = Field(default=None, description="Context manager module")


class ResolvedSessionConfig(CamelCaseModel):
    """Session configuration with source tracking."""

    orchestrator: ResolvedModuleRef | None = Field(default=None, description="Orchestrator module with source")
    context: ResolvedModuleRef | None = Field(default=None, description="Context manager module with source")


class BundleListItem(CamelCaseModel):
    """Bundle list item with summary information."""

    name: str = Field(description="Bundle name")
    version: str = Field(default="1.0.0", description="Bundle version")
    description: str | None = Field(default=None, description="Bundle description")
    source: str = Field(description="Bundle source: 'user' or 'system'")
    path: str = Field(description="Path to bundle file")

    # Quick stats
    provider_count: int = Field(default=0, description="Number of providers")
    tool_count: int = Field(default=0, description="Number of tools")
    hook_count: int = Field(default=0, description="Number of hooks")
    agent_count: int = Field(default=0, description="Number of agents")

    # Composition
    includes: list[str] = Field(default_factory=list, description="Included bundle references")


class BundleDetails(CamelCaseModel):
    """Full bundle details (raw structure)."""

    name: str = Field(description="Bundle name")
    version: str = Field(default="1.0.0", description="Bundle version")
    description: str | None = Field(default=None, description="Bundle description")
    source: str = Field(description="Bundle source: 'user' or 'system'")
    path: str = Field(description="Path to bundle file")

    # Quick stats
    provider_count: int = Field(default=0, description="Number of providers")
    tool_count: int = Field(default=0, description="Number of tools")
    hook_count: int = Field(default=0, description="Number of hooks")
    agent_count: int = Field(default=0, description="Number of agents")

    # Composition
    includes: list[str] = Field(default_factory=list, description="Included bundle references")

    # Full structure
    session: SessionConfigRef | None = Field(default=None, description="Session configuration")
    providers: list[ModuleRef] = Field(default_factory=list, description="Provider modules")
    tools: list[ModuleRef] = Field(default_factory=list, description="Tool modules")
    hooks: list[ModuleRef] = Field(default_factory=list, description="Hook modules")
    agents: list[ModuleRef] = Field(default_factory=list, description="Agent modules")
    context: dict[str, str] = Field(default_factory=dict, description="Context file mappings")
    instruction: str | None = Field(default=None, description="System instruction (markdown body)")


class ResolvedBundle(CamelCaseModel):
    """Flattened bundle with source tracking.

    This is the resolved view showing what will actually run,
    with tracking of which bundle contributed each component.
    """

    name: str = Field(description="Bundle name")
    source: str = Field(description="Bundle source: 'user' or 'system'")
    includes_chain: list[str] = Field(
        default_factory=list, description="Full includes chain (e.g., ['foundation/base', 'basic', 'my-bundle'])"
    )

    # Resolved sections with source tracking
    session: ResolvedSessionConfig | None = Field(default=None, description="Session configuration with source")
    providers: list[ResolvedModuleRef] = Field(default_factory=list, description="Provider modules with source")
    tools: list[ResolvedModuleRef] = Field(default_factory=list, description="Tool modules with source")
    hooks: list[ResolvedModuleRef] = Field(default_factory=list, description="Hook modules with source")
    agents: list[ResolvedModuleRef] = Field(default_factory=list, description="Agent modules with source")

    # Content from final bundle
    instruction: str | None = Field(default=None, description="System instruction")


class BundleSource(CamelCaseModel):
    """Raw bundle source content."""

    name: str = Field(description="Bundle name")
    content: str = Field(description="Raw bundle file content")
    path: str = Field(description="Path to bundle file")
    format: str = Field(description="File format: 'md', 'yaml', or 'directory'")


class CreateBundleRequest(CamelCaseModel):
    """Request to create a new bundle."""

    name: str = Field(pattern=r"^[a-z0-9-]+$", description="Bundle name (kebab-case)")
    base_bundle: str | None = Field(default=None, description="Bundle to extend (creates include)")
    description: str | None = Field(default=None, description="Bundle description")


class CopyBundleRequest(CamelCaseModel):
    """Request to copy a bundle."""

    new_name: str = Field(pattern=r"^[a-z0-9-]+$", description="New bundle name (kebab-case)")


class UpdateBundleRequest(CamelCaseModel):
    """Request to update a bundle."""

    content: str = Field(description="Raw bundle content (markdown with YAML frontmatter)")
